docs/container.md

```diff
@@ -0,0 +1,313 @@
+        $bar = // ...
+
+        return Foo::createSomething($bar);
+    }),
+);
+```
+
+#### Configuring containers in tests
+
+When writing integration or system tests you can inject your own classes (such as mocks) into the Piwik environment one of two ways:
+
+1. Add configuration to your plugin's `config/test.php` file. For example:
+
+  ```php
+  <?php
+
+  return array(
+      'Piwik\Plugins\MyPlugin\MyRESTClient' => DI\object('Piwik\Plugins\MyPlugin\Test\MockRESTClient'),
+  );
+  ```
+
+  This configuration is applied in every test, however, so it may not be desirable.
+
+2. Override the `provideContainerConfig()` method in either your test case class or your `Fixture` class and return the DI config. For example:
+
+  ```php
+  use Piwik\Tests\Framework\TestCase\IntegrationTestCase;
+
+  class APITest extends IntegrationTestCase
+  {
+      // ... test cases ...
+
+      public function provideContainerConfig()
+      {
+          return array(
+              'Piwik\Plugins\MyPlugin\Dao\MyEntityDao' => DI\object('Piwik\Plugins\MyPlugin\Test\Mock\MockMyEntityDao')
+                  ->constructorParameter('tmpPath', '/my/test/tmp/path'),
+          );
+      }
+  }
+  ```
+
+Both of these types of container configuration affect child processes as well. So, for example, tracker requests sent in Fixtures will use this overridden configuration.
+
+**Configuring the container in UI tests**
+
+For UI tests which are written in JavaScript, this is a bit trickier. You can either use the first approach above, or override the DI config in a Fixture class and use it in your tests.
+
+For example, in your Fixture:
+
+```php
+use Piwik\Tests\Framework\Fixture;
+
+class MyFixture extends Fixture
+{
+    // ...
+
+    public function provideContainerConfig()
+    {
+        // ...
+    }
+}
+```
+
+then in your UI test:
+
+```js
+describe("MyUiTest", function () {
+    this.fixture = "Piwik\\Plugins\\MyPlugin\\Test\\Fixtures\\MyFixture";
+
+    // ... tests ...
+})
+```
+
+## More about Dependency Injection in Piwik
+
+This section describes 
+
+The **DI container** is used to store and configure all of Piwik's application logic. That is to say, any class that encapsulates logic important to Piwik and Piwik only should be stored in the container.
+
+We use dependency injection for the following reasons:
+
+* to move the responsibility of resolving dependencies outside of individual classes. Classes only list what they need, and the DI container gives it to them.
+* to allow plugins and users the ability to extend Piwik in any way they need by replacing and extending objects in the DI container.
+* to allow developers to more easily test classes both in isolation and in the Piwik environment, by making it easy to inject mocks.
+* to allow developers to more easily maintain code modularity by avoiding hard-coded dependencies.
+
+### What classes go in the container
+
+In Piwik development, there are four types of classes that exist:
```

Comment:
Is this part kinda repeating the first part? Maybe also rather something for  a page under http://developer.piwik.org/piwik-in-depth ?